src/libstore/build/derivation-goal.cc

```diff
@@ -2097,15 +2094,15 @@ struct RestrictedStore : public LocalFSStore, public virtual RestrictedStoreConf
 
     void registerDrvOutput(const Realisation & info) override
     {
-        if (!goal.isAllowed(info.id.drvPath))
-            throw InvalidPath("cannot register unknown drv output '%s' in recursive Nix", printStorePath(info.id.drvPath));
+        // XXX: Should we check for something here? Probably, but I'm not sure
+        // how
         next->registerDrvOutput(info);
     }
 
     std::optional<const Realisation> queryRealisation(const DrvOutput & id) override
     {
-        if (!goal.isAllowed(id.drvPath))
-            throw InvalidPath("cannot query the output info for unknown derivation '%s' in recursive Nix", printStorePath(id.drvPath));
+        // XXX: Should we check for something here? Probably, but I'm not sure
+        // how
```

Comment:
`queryRealisation()` should only succeed in `RestrictedStore` if `DrvOutput` refers to a derivation that the client added (via `addToStore`) *and* where it has built the output (i.e. has done a `buildDerivation()`). Otherwise it can leak info to which the sandboxed build isn't supposed to have access.

Reply:
Indeed… like https://github.com/NixOS/nix/pull/4348#discussion_r541031277 I'll just restrict everything for now as it's slightly tricky to implement and we might change it eventually anyways.

I've opened #4353 to track it